x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java

```diff
@@ -660,62 +658,53 @@ public List<NamedExpression> visitGrouping(EsqlBaseParser.FieldsContext ctx) {
-            return new Literal(source, null, dataType);
-        }
-        final DataType sourceType;
-        try {
-            sourceType = DataTypes.fromJava(param.value);
-        } catch (QlIllegalArgumentException ex) {
-            throw new ParsingException(
-                ex,
-                source,
-                "Unexpected actual parameter type [{}] for type [{}]",
-                param.value.getClass().getName(),
-                param.type
-            );
-        }
-        if (sourceType == dataType) {
-            // no conversion is required if the value is already have correct type
-            return new Literal(source, param.value, dataType);
-        }
-        // otherwise we need to make sure that xcontent-serialized value is converted to the correct type
-        try {
+        QueryParam param = paramByToken(ctx.PARAM());
+        return visitParam(ctx, param);
+    }
 
-            if (EsqlDataTypeConverter.canConvert(sourceType, dataType) == false) {
-                throw new ParsingException(
-                    source,
-                    "Cannot cast value [{}] of type [{}] to parameter type [{}]",
-                    param.value,
-                    sourceType,
-                    dataType
-                );
-            }
-            return new Literal(source, EsqlDataTypeConverter.converterFor(sourceType, dataType).convert(param.value), dataType);
-        } catch (QlIllegalArgumentException ex) {
-            throw new ParsingException(ex, source, "Unexpected actual parameter type [{}] for type [{}]", sourceType, param.type);
+    @Override
+    public Object visitInputNamedOrPositionalParam(EsqlBaseParser.InputNamedOrPositionalParamContext ctx) {
+        QueryParam param = paramByNameOrPosition(ctx.NAMED_OR_POSITIONAL_PARAM());
+        return visitParam(ctx, param);
+    }
+
+    private Object visitParam(EsqlBaseParser.ParamsContext ctx, QueryParam param) {
+        Source source = source(ctx);
+        DataType type = param.type();
+        if (type == null) {
+            throw new ParsingException(source, "Invalid parameter data type [{}]", type);
         }
+        return new Literal(source, param.value(), type);
     }
 
-    private TypedParamValue param(TerminalNode node) {
+    QueryParam paramByToken(TerminalNode node) {
         if (node == null) {
             return null;
         }
-
         Token token = node.getSymbol();
-
-        if (params.containsKey(token) == false) {
+        if (params.contains(token) == false) {
             throw new ParsingException(source(node), "Unexpected parameter");
         }
-
         return params.get(token);
     }
 
+    QueryParam paramByNameOrPosition(TerminalNode node) {
+        if (node == null) {
+            return null;
+        }
+        Token token = node.getSymbol();
+        String nameOrPosition = token.getText().substring(1);
+        if (nameOrPosition.chars().allMatch(Character::isDigit)) {
```

Comment:
Nit: if there's no method in the utils to do the check, add a simple static method that uses for/charAt instead of using streams. They do work but are inefficient.